Fix riscv32 atomics and fix tests on 32-bit platforms

xobs opened the pull request:
This fixes tests on 32-bit platforms. Previously, the `Encode` would encode a `usize` and decode a `u64`. This works fine on 64-bit platforms, but on 32-bit platforms this results in a mismatch.

Force `usize` to be encoded as a `u64` even on 32-bit platforms.

Additionally, the 32-bit riscv target does not implement `AtomicU64` or `AtomicI64`, so remove those impelementations on that platform.

xobs committed:
tests: fix alloc range test on 32-bit Windows

This test checks the range, which is necessarily different when running
on another platform.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
enc: case usize/isize as u64/i64 in serialization

The deserialization process assumes that usize/isize are 64-bits, as
does the spec at
https://github.com/bincode-org/bincode/blob/trunk/docs/spec.md#varintencoding

Force `usize` and `isize` to be encoded as `u64` and `i64` to force
32-bit platforms to conform to this spec.

This fixes running tests under `cargo test --target i686-pc-windows-msvc`

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
atomic: only provide Atomic*64 on supported platforms

Not all platforms support AtomicI64 or AtomicU64. Use the
`target_has_atomic` config flag to determine whether to include these.

This fixes #532.

Signed-off-by: Sean Cross <sean@xobs.io>

VictorKoenders commented on src/features/atomic.rs:
```diff
@@ -64,6 +67,7 @@ impl Decode for AtomicU32 {
     }
 }
 
+#[cfg(not(target_arch = "riscv32"))]
```
This should also be `target_has_atomic`

VictorKoenders commented on src/features/atomic.rs:
```diff
@@ -73,6 +77,7 @@ impl Encode for AtomicU64 {
     }
 }
 
+#[cfg(not(target_arch = "riscv32"))]
```
This should also be `target_has_atomic`

VictorKoenders reviewed (commented):
I wonder if we could make this bigger and add `target_has_atomic` to all atomic impls, and deprecate the `atomic` feature

xobs committed:
atomic: remove feature and use new attributes

Now that there is an attribute to indicate which atomic features are
supported on this platform, remove the `atomic` Feature and use these
new attributes.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
Thanks for the suggestions. I've annotated all Atomic implementations.

I assume that `Bool` is the same as `8`, and `usize` is the same as `ptr`.

I'm not sure what the best way is to deprecate the feature, so I removed it wholesale.

VictorKoenders commented:
Can you pull in the new cross-platform build pipeline, and see if bincode now compiles for your target arch?

xobs committed:
Merge remote-tracking branch 'upstream/trunk' into fix-tests-32-bit

xobs commented:
It does build for my platform still, so it looks good to me!

```sh
[10:10:39 am] E:/Code/bincode> cargo build --lib --target riscv32imac-unknown-xous-elf
   Compiling virtue v0.0.7
   Compiling bincode_derive v2.0.0-rc.1 (E:\Code\bincode\derive)
   Compiling bincode v2.0.0-rc.1 (E:\Code\bincode)
    Finished dev [unoptimized + debuginfo] target(s) in 2.39s
[10:10:43 am] E:/Code/bincode>
```

xobs committed:
alloc: run `cargo fmt`

Signed-off-by: Sean Cross <sean@xobs.io>

VictorKoenders commented:
If you run `cargo fmt` and `cargo clippy` I think this is good to merge

xobs commented:
Alright, I just ran `cargo fmt` to fix that up.

VictorKoenders commented:
Thanks!
